sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedIterable.java

```diff
@@ -187,17 +194,48 @@ public PagedIterable(Function<Integer, PagedResponse<T>> firstPageRetriever,
      */
     @SuppressWarnings("deprecation")
     public <S> PagedIterable<S> mapPage(Function<T, S> mapper) {
-        return new PagedIterable<>(pagedFlux.mapPage(mapper));
+        if (pagedFlux != null) {
+            return new PagedIterable<>(pagedFlux.mapPage(mapper));
+        }
+        Function<Integer, PagedResponse<S>> firstMappedPageRetriever = pageSize -> {
+            PagedResponse<T> firstPageResponse = this.firstPageRetriever.apply(pageSize);
+            PagedResponse<S> firstMappedPageResponse = mapPagedResponse(firstPageResponse, mapper);
+            return firstMappedPageResponse;
+        };
+        BiFunction<String, Integer, PagedResponse<S>> nextMappedPageRetriever = (continuationToken, pageSize) -> {
+            PagedResponse<T> nextPageResponse = this.nextPageRetriever.apply(continuationToken, pageSize);
+            PagedResponse<S> nextMappedPageResponse = mapPagedResponse(nextPageResponse, mapper);
+            return nextMappedPageResponse;
+        };
+        return new PagedIterable<>(firstMappedPageRetriever, nextMappedPageRetriever);
     }
 
     /**
      * Create PagedIterable backed by Page Retriever Function Supplier.
      *
-     * @param provider the Page Retrieval Provider
-     * @param ignored param is ignored, exists in signature only to avoid conflict with first ctr
+     * @param provider the Page Retrieval Provider.
+     * @param firstPageRetriever first page retriever function to get the first page given the page size.
+     * @param nextPageRetriever next page retriever function to get the next page given a continuation token and the page size.
      */
-    private PagedIterable(Supplier<PageRetrieverSync<String, PagedResponse<T>>> provider, boolean ignored) {
+    private PagedIterable(Supplier<PageRetrieverSync<String, PagedResponse<T>>> provider,
+                          Function<Integer, PagedResponse<T>> firstPageRetriever,
+                          BiFunction<String, Integer, PagedResponse<T>> nextPageRetriever) {
         super(provider);
         this.pagedFlux = null;
+        this.firstPageRetriever = firstPageRetriever;
+        this.nextPageRetriever = nextPageRetriever;
     }
+
+    private <S> PagedResponse<S> mapPagedResponse(PagedResponse<T> pagedResponse, Function<T, S> mapper) {
+        if (pagedResponse == null) {
+            return null;
+        }
+        return new PagedResponseBase<String, S>(pagedResponse.getRequest(),
+            pagedResponse.getStatusCode(),
+            pagedResponse.getHeaders(),
+            pagedResponse.getValue().stream().map(mapper).collect(Collectors.toList()),
```

Comment:
Original comment by @alzimmermsft about using new ArrayList() instead https://github.com/Azure/azure-sdk-for-java/pull/34905#discussion_r1196407150